### DISCUSSION TITLE
Problem when adding a language server to the Cairo language

### DISCUSSION THREAD

**Discussion by Eikix** [helix-editor/helix#10250] [Troubleshooting]
Hi all:)!

I've recently picked up Helix and love it! Thanks a lot. Been using it as my main Rust IDE 🦀.

I'm currently working in the Starknet ecosystem and their programming language is [Cairo](https://www.cairo-lang.org/cairo-whitepaper/). 
I saw that in the default language config, [Cairo is already mentioned](https://github.com/helix-editor/helix/blob/master/languages.toml#L1523). Nonetheless, there is no language server associated.

I tried to override the basic config with the following language config (`~/.config/helix/languages.toml`):
```
[[language]]
name = "cairo"
scope = "source.cairo"
injection-regex = "cairo"
file-types = ["cairo"]
indent = { tab-width = 2, unit = "  " }
comment-token = "#"
language-server = { command = "cairo-ls" }
```

Where I aliased `cairo-ls` in my `.zshrc` to be:
`alias cairo-ls="node ~/code/lsp/node_modules/cairo-ls/out/server.js --stdio"`

As the cairo language server does not have a binary but rather a node script: 
https://www.npmjs.com/package/cairo-ls

When trying to open a `.cairo` file, Helix informs me that there is no language server running.

**Comment by the-mikedavis** [comment]
We spawn the language server directly as a child process instead of using a shell so the `language-server` configuration won't find commands from shell aliases. It should work to create a wrapper script that executes that command and to add that to your `$PATH`, or you can configure that `node` command:

```toml
# ~/.config/helix/languages.toml
[[language]]
name = "cairo"
language-server = { command = "node", args = ["~/code/lsp/node_modules/cairo-ls/out/server.js", "--stdio"] }
```

**Comment by Eikix** [comment]
Creating a .sh wrapper script and adding it to my path works fine! 
Thanks for your help!

PS: I tried the second option, it didn't work for me


Thanks again for your fast response!! 🥷🔥

**Comment by Eikix** [comment]
Hey!
Although charging/loading the LSP works, Helix does not recognise where to look for relative imports in the Cairo syntax.

<img width="1691" alt="image" src="https://user-images.githubusercontent.com/66871571/208955570-922f65f1-8945-4245-9284-1f1481422418.png">

How can I make the LSP recognise correct paths to check: `src/`, `tests/`, `libs/`, etc?

**Comment by archseer** [comment]
You probably either need to pass more flags to the language server, or add a configuration section.

**Comment by archseer** [comment]
I previously tested out cairo-ls but found it too minimal so I personally don't use it anymore

**Comment by Eikix** [comment]
> I previously tested out cairo-ls but found it too minimal so I personally don't use it anymore

What alternatives are there;)?

**Comment by Eikix** [comment]
> You probably either need to pass more flags to the language server, or add a configuration section.

What is a configuration section? Is that a helix thing?

**Comment by the-mikedavis** [comment]
You can control what configuration is sent to the server in the initialization function with the `config` key in language configuration: https://docs.helix-editor.com/master/languages.html#language-server-configuration

For an example see the deno configurations: https://github.com/helix-editor/helix/wiki/How-to-install-the-default-language-servers#deno

**Comment by Eikix** [comment]
Thanks for your time:)! I've checked it out,

Tried to pass
`config = { sourceDir = "tests", venvCommand = "source ~/cairo_venv/bin/activate", nileUseVenv = false, useVenv = true, nileVenvCommand = "source ~/cairo_venv/bin/activate" }`

In order to try and reproduce the VSCode equivalent (in `settings.json`):
```json
{
  "cairols.nileUseVenv": false,
  "cairols.venvCommand": "source ~/cairo_venv/bin/activate",
  "cairols.useVenv": true,
  "cairols.cairoPath": ["lib", "tests"],
  "cairols.sourceDir": "src",
  "cairols.nileVenvCommand": "source ~/cairo_venv/bin/activate"
}
```

It does not work. I still get an error when trying to import modules inside a .cairo file:

<img width="727" alt="image" src="https://user-images.githubusercontent.com/66871571/209229754-1abfc823-430b-4980-a25a-6fa0ba5c67f4.png">
Do you have an idea why it does not work?

**Comment by archseer** [comment]
Provide logs after running hx with the -v flag. I guessing you need to provide a full path rather than using `~`

**Comment by Eikix** [comment]
It appears the cairo language server searched imports first in `Users/eliastazartes/code/theodo/kakarot/kakarot/constants.cairo`

But it would find it in:
`Users/eliastazartes/code/theodo/kakarot/src/kakarot/constants.cairo`

This is my main issue as of now, not sure about the `venvCommand` and the `~` symbol to express the homeDir.


In the [cairo-ls library's README](https://github.com/ericglau/cairo-ls/tree/main), it is written:
> If your project's contracts directory is in a subfolder such as src, set the folder name using the sourceDir setting in your IDE or language client settings.

How can I change the language client's settings in Helix?

**Comment by the-mikedavis** [comment]
I think they're talking about the config sent in the initialization function - the `config` key in `languages.toml`

**Comment by Eikix** [comment]
```toml
[[language]]
name = "cairo"
scope = "source.cairo"
injection-regex = "cairo"
file-types = ["cairo"]
indent = { tab-width = 2, unit = "  " }
comment-token = "#"
config = { sourceDir = "tests", venvCommand = "source ~/cairo_venv/bin/activate", nileUseVenv = false, useVenv = true, nileVenvCommand = "source ~/cairo_venv/bin/activate" }
language-server = { command = "cairols.sh" }
```

This is my language config in languages.toml

**Comment by gabydd** [comment]
Have no idea if this is correct but the vscode config has ``sourceDir`` as ``src`` whereas the helix one has it as ``tests`` so maybe change that to ``src``.

**Comment by Eikix** [comment]
> Have no idea if this is correct but the vscode config has ``sourceDir`` as ``src`` whereas the helix one has it as ``tests`` so maybe change that to ``src``. 

Ah yes, I was playing around with the config object but I tried `src` too without success, and all the other root folders of the project.
